qiskit/circuit/instructionset.py

```diff
@@ -58,17 +162,34 @@ def inverse(self):
             self.instructions[index] = instruction.inverse()
         return self
 
-    def c_if(self, classical, val):
-        """Add condition on classical register to all instructions."""
-        if self.circuit_cregs:
-            circuit_clbits = [bit for creg in self.circuit_cregs for bit in creg]
-        if isinstance(classical, (Clbit, int)):
-            if isinstance(classical, int):
-                classical = circuit_clbits[classical]
-            for creg in self.circuit_cregs:
-                if creg[:] == [classical]:
-                    classical = creg
-                    break
+    def c_if(self, classical: Union[Clbit, ClassicalRegister, int], val: int) -> "InstructionSet":
+        """Set a classical equality condition on all the instructions in this set between the
+        :obj:`.ClassicalRegister` or :obj:`.Clbit` ``classical`` and value ``val``.
+
+        .. note::
+            This overrides any previously set condition.
+
+        Args:
+            classical: the classical resource the equality condition should be on.  If this is given
+                as an integer, it will be resolved into a :obj:`.Clbit` using the same conventions
+                as the circuit these instructions are attached to.
+            val: the value the classical resource should be equal to.
+
+        Returns:
+            This same instance of :obj:`.InstructionSet`, but now mutated to have the given equality
+            condition.
+
+        Raises:
+            CircuitError: if the passed classical resource is invalid, or otherwise not resolvable
+                to a concrete resource that these instructions are permitted to access.
+        """
+        if self._resolver is None and not isinstance(classical, (Clbit, ClassicalRegister)):
+            raise CircuitError(
+                "Cannot pass an index as a condition variable without specifying a resolver"
+                " when creating this InstructionSet."
+            )
+        if self._resolver is not None:
+            classical = self._resolver(classical)
         for gate in self.instructions:
             gate.c_if(classical, val)
         return self
```

Comment:
It is a bit weird if we don't allow chaining here that we return `self` like this.

Reply:
Yeah, seems slightly odd. Perhaps we were thinking of adding more methods to `Instruction`/`InstructionSet` so you may want to do multiple things at once.